### PR TITLE
fix cmake install config package

### PR DESCRIPTION
**Pull request by ClausKlein**
use CPM.cmake to install the project
cleanup cmake file
format cmake file
fix compile errors
fix some clang-tidy warnings
add some Notes to source

Problem:
- dit not compile on OSX with gcc-10, clang-11, AppleClangV12

Solution:
- use target_link_library()
- and change some compiler flags
- correct source code

Issue: this fix https://github.com/boost-ext/ut/issues/420 too

Reviewers:
@

### CODE REVIEW

**Comment by krzysztof-jusiak** [comment]
Thanks, @ClausKlein for working it, I really appreciate it :pray:

**Comment by ClausKlein** [comment]
> Thanks, @ClausKlein for working it, I really appreciate it 🙏

I can't understand what goes wrong here
https://travis-ci.org/github/boost-ext/ut/jobs/760048262#L109

**Comment by ClausKlein** [comment]
@krzysztof-jusiak would you like to migrate to GitHub workflows?

see https://github.com/ClausKlein/ut/actions

**Comment by krzysztof-jusiak** [comment]
Switching to github actions would be great, thank you @ClausKlein :pray:

File: CMakeLists.txt [hunk 1]
```diff
@@ -4,165 +4,163 @@
 # Distributed under the Boost Software License, Version 1.0.
 # (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 #
-cmake_minimum_required(VERSION 3.1...3.19)
-project(boost.ut CXX)
-
-include(CTest)
+cmake_minimum_required(VERSION 3.12...3.20)
+project(
+  ut
```
Comment (ClausKlein) [inline CMakeLists.txt:9]: the project name is changed, OK?

File: CMakeLists.txt [hunk 2]
```diff
@@ -4,165 +4,163 @@
...
+
+  # PackageProject.cmake will be used to make our target installable
+  CPMAddPackage(
+    NAME PackageProject.cmake
+    GITHUB_REPOSITORY TheLartians/PackageProject.cmake
+    VERSION 1.6.0
+  )
+endif()
+
+add_library(ut INTERFACE)
```
Comment (ClausKlein) [inline CMakeLists.txt:52]: the lib name is ut, same as project!

File: CMakeLists.txt [hunk 3]
```diff
@@ -4,165 +4,163 @@
...
+#
+# /usr/local/include/boost -> ../Cellar/boost/1.75.0_1/include/boost
+# TODO: set(INCLUDE_INSTALL_DIR boost-${BOOST_VERSION}/include)
+
+set(INCLUDE_INSTALL_DIR include/${PROJECT_NAME}-${PROJECT_VERSION}/include)
+# XXX variant: set(INCLUDE_INSTALL_DIR include)
+target_include_directories(ut INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include> $<INSTALL_INTERFACE:${INCLUDE_INSTALL_DIR}>)
+if(APPLE)
+  target_compile_features(ut INTERFACE cxx_std_20)
+  # else() target_compile_features(ut INTERFACE cxx_std_17)
```
Comment (ClausKlein) [inline CMakeLists.txt:65]: what is the min required std?

File: CMakeLists.txt [hunk 4]
```diff
@@ -4,165 +4,163 @@
...
+endif()
+
+if("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
+  if(WIN32) # clang-cl
+    # FIXME: we should not export this pedantic options! CK
```
Comment (ClausKlein) [inline CMakeLists.txt:70]: clang-cl is buggy for me?

File: CMakeLists.txt [hunk 5]
```diff
@@ -4,165 +4,163 @@
...
-                        -Wno-c++98-compat-pedantic
-                        -Wno-c99-extensions
-                        -Wno-global-constructors
-                        -Wno-exit-time-destructors
-                        -Wno-missing-variable-declarations
-                        -Wno-class-varargs
-                        -Wno-padded
-                        -Wno-missing-prototypes
-                        -Wno-ctad-maybe-unsupported)
+    add_compile_options(-Wall -Wextra -Wpedantic -Werror)
```
Comment (ClausKlein) [inline CMakeLists.txt:84]: that should be good for all non windows compiler

File: CMakeLists.txt [hunk 6]
```diff
@@ -4,165 +4,163 @@
...
-                      -Wduplicated-branches
-                      -Wcast-align=strict
-                      -Wmissing-include-dirs
-                      -Wno-missing-declarations)
-elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
-  target_compile_options(boost.ut INTERFACE
-                      /W4
-                      /WX)
+elseif("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU")
+  # FIXME: we should not export this pedantic options! CK
```
Comment (ClausKlein) [inline CMakeLists.txt:87]: Note this!

File: CMakeLists.txt [hunk 7]
```diff
@@ -4,165 +4,163 @@
...
-          "${multiValues}"
-          ${ARGN})
-
-  if (BOOST_UT_ENABLE_RUN_AFTER_BUILD)
-    add_custom_command(TARGET ${PARSE_TARGET} COMMAND ${PARSE_COMMAND})
-  else()
-    add_test(NAME ${PARSE_TARGET} COMMAND ${PARSE_COMMAND})
-  endif()
-endfunction()
+# Create imported target Boost::ut
```
Comment (ClausKlein) [inline CMakeLists.txt:139]: the exported interface lib name

File: example/CMakeLists.txt [hunk 8]
```diff
@@ -22,18 +30,22 @@ endfunction()
...
-  tbb_build(TBB_ROOT ${TBB_ROOT} CONFIG_DIR TBB_DIR MAKE_ARGS stdver="c++17")
-  find_package(TBB)
-  target_link_libraries(boost_ut_parallel_runner ${TBB_IMPORTED_TARGETS})
+if(APPLE OR "${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
+  example(cfg/parallel_runner parallel_runner)
 endif()
 
+# if(DEFINED ENV{TBB_ROOT})
+#   set(TBB_ROOT $ENV{TBB_ROOT})
+#   # FIXME: include could not find load file:
```
Comment (ClausKlein) [inline example/CMakeLists.txt:39]: Note this, it does not work at all?

File: example/section.cpp [hunk 9]
```diff
@@ -33,15 +33,24 @@ int main() {
...
+      mut(str) << '1';
+      expect(str.str() == "1"sv);
+    };
+  }
 
-  "[str2]"_test = [str{std::move(str)}] {
-    mut(str) << '2';
-    expect(str.str() == "2"sv);
-  };
+  // FIXME: [clang-analyzer-cplusplus.Move,-warnings-as-errors]
```
Comment (ClausKlein) [inline example/section.cpp:45]: this was a real error!

File: test/CMakeLists.txt [hunk 10]
```diff
@@ -4,5 +4,21 @@
 # Distributed under the Boost Software License, Version 1.0.
 # (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 #
+cmake_minimum_required(VERSION 3.12...3.20)
+project(boost.ut.test LANGUAGES CXX)
+
+if(TEST_INSTALLED_VERSION)
+  if(NOT DEFINED CMAKE_CXX_STANDARD)
+    option(CXX_STANDARD_REQUIRED "Require c++ standard" YES)
```
Comment (ClausKlein) [inline test/CMakeLists.txt:12]: is this default OK?

File: test/ut/ut.cpp [hunk 11]
```diff
@@ -338,7 +338,8 @@ int main() {
       static_assert(42u == 42_ul);
       static_assert(42.42f == 42.42_f);
       static_assert(42.42 == 42.42_d);
-      static_assert(240.209996948 == 240.209996948);
+      static_assert(240.209996948 ==
+                    240.209996948);  // FIXME: g++-10 [-Werror=float-equal]
```
Comment (ClausKlein) [inline test/ut/ut.cpp:342]: note this

File: CMakeLists.txt (hunk 3 above)
Comment (krzysztof-jusiak) [reply CMakeLists.txt:65]: c++17/20 I'd say

File: CMakeLists.txt (hunk 4 above)
Comment (krzysztof-jusiak) [reply CMakeLists.txt:70]: not sure :thinking:

File: example/CMakeLists.txt (hunk 8 above)
Comment (krzysztof-jusiak) [reply example/CMakeLists.txt:39]: that used to work on travis but somehow got broken recently, not sure why

File: example/section.cpp (hunk 9 above)
Comment (krzysztof-jusiak) [reply example/section.cpp:45]: do you have a fix or what me to add one?

File: test/ut/ut.cpp (hunk 11 above)
Comment (krzysztof-jusiak) [reply test/ut/ut.cpp:342]: I'll take a look at this one :point_up:

File: example/section.cpp (hunk 9 above)
Comment (ClausKlein) [reply example/section.cpp:45]: it is fixed witch my patch

File: CMakeLists.txt (hunk 3 above)
Comment (ClausKlein) [reply CMakeLists.txt]: ok

File: CMakeLists.txt (hunk 4 above)
Comment (ClausKlein) [reply CMakeLists.txt:65]: hundreds of warnings is not normal?

Comment (krzysztof-jusiak) [reply CMakeLists.txt:65]: I hope not

**Comment by krzysztof-jusiak** [comment]
I'm happy to merge it to carry one if you want @ClausKlein, could we also process with github actions?

File: CMakeLists.txt (hunk 4 above)
Comment (ClausKlein) [reply CMakeLists.txt:65]: see https://github.com/ClausKlein/ut/runs/1981508898?check_suite_focus=true

**Comment by ClausKlein** [comment]
merge it please, than we can continue with more GitHub actions ...

**Comment by krzysztof-jusiak** [comment]
Merged @ClausKlein, thanks again and super exciting about the follow-up with github actions :pray:

**Comment by ClausKlein** [comment]
Under settings you should now enable the GitHub actions

**Comment by krzysztof-jusiak** [comment]
done :point_up:

**Pull request merged**